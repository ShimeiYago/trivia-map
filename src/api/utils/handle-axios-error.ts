import { AxiosError } from 'axios';

export function handleAxiosError(axiosError: AxiosError): ApiError {
  if (axiosError.response) {
    // when API is active but response includes error
    const error: ApiError = {
      status: axiosError.response.status,
      data: axiosError.response.data,
      errorMsg: axiosError.message,
    };
    return error;
  } else {
    // when API is not active
    const error: ApiError = {
      status: 500,
      data: {},
      errorMsg: axiosError.message,
    };
    return error;
  }
}

export type ApiError = {
  status: number;
  data: unknown;
  errorMsg: string;
};
