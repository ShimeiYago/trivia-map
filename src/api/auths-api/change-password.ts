import { BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function changePassword(
  password1: string,
  password2: string,
): Promise<void> {
  const axiosInstance = getAxiosInstance({}, {});

  const requestData: ChangePasswordRequest = {
    password1: password1,
    password2: password2,
  };

  try {
    // TODO: Set reasonable timeout
    await axiosInstance.post(`${BASE_URL}/auths/password/change`, requestData);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type ChangePasswordRequest = {
  password1: string;
  password2: string;
};

export type ValidationError = {
  password1?: string[];
  password2?: string[];
};
