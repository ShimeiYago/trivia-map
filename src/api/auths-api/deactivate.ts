import { API_TIMEOUT, BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function deactivate(): Promise<void> {
  const axiosInstance = getAxiosInstance({ timeout: API_TIMEOUT.long }, {});

  try {
    await axiosInstance.put(`${BASE_URL}/auths/deactivate/`);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
