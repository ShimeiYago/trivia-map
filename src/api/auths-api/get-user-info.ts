import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { User } from 'types/user';
import { mockGetUserInfoResponse } from 'api/mock/auths-response/get-user-info';

export async function getUserInfo(): Promise<User> {
  const axiosInstance = getAxiosInstance({}, mockGetUserInfoResponse);

  try {
    // TODO: Set reasonable timeout
    const res: AxiosResponse<User> = await axiosInstance.get(`${BASE_URL}/auths/user`);
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
