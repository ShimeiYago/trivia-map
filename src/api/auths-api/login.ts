import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { mockLoginResponse } from 'api/mock/auths-response/login';
import { User } from 'types/user';

export async function login(
  email: string,
  password: string,
): Promise<LoginResponse> {
  const axiosInstance = getAxiosInstance({}, mockLoginResponse);

  const requestData: LoginRequest = {
    email: email,
    password: password,
  };

  try {
    // TODO: Set reasonable timeout
    const res: AxiosResponse<LoginResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/login`,
      requestData,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type LoginRequest = {
  email: string;
  password: string;
};

export type LoginResponse = {
  access_token: string;
  refresh_token: string;
  user: User;
};

export type ValidationError = {
  email?: string[];
  password?: string[];
  non_field_errors?: string[];
};
