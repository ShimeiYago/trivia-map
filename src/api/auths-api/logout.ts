import { BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function logout(): Promise<void> {
  const axiosInstance = getAxiosInstance({}, {});

  try {
    await axiosInstance.post(`${BASE_URL}/auths/logout`);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
