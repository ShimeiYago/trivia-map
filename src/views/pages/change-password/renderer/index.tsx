import React from 'react';
import { LoadingState } from 'types/loading-state';
import { Alert, Box, Stack, TextField, Typography } from '@mui/material';
import { LoadingButton } from '@mui/lab';
import { HeaderErrorMessages } from 'views/components/moleculars/header-error-messages';
import { changePassword, ValidationError } from 'api/auths-api/change-password';
import { ApiError } from 'api/utils/handle-axios-error';
import { globalAPIErrorMessage } from 'constant/global-api-error-message';
import { AdminWrapper } from 'views/components/organisms/admin-wrapper';
import { autoRefreshApiWrapper } from 'utils/auto-refresh-api-wrapper';
import { BackToNavi } from 'views/components/moleculars/back-to-navi';
import { ACCOUNT_SETTINGS_LINK } from 'constant/links';
import { PAGE_NAMES } from 'constant/page-names';
import { User } from 'types/user';

export class Renderer extends React.Component<Props, State> {
  state: State = {
    password1: '',
    password2: '',
    loadingState: 'waiting',
  };

  render() {
    const disabled = this.state.loadingState === 'loading' || this.props.user?.isSocialAccount;

    const form = (
      <Box component="form" noValidate>
        <TextField
          margin="normal"
          fullWidth
          name="password1"
          label="パスワード"
          type="password"
          id="password1"
          disabled={disabled}
          helperText={this.state.formError?.password1}
          error={!!this.state.formError?.password1}
          onChange={this.handleChangeTextField('password1')}
        />
        <TextField
          margin="normal"
          fullWidth
          name="password2"
          label="パスワード（確認）"
          type="password"
          id="password2"
          disabled={disabled}
          helperText={this.state.formError?.password2}
          error={!!this.state.formError?.password2}
          onChange={this.handleChangeTextField('password2')}
        />
        <LoadingButton
          fullWidth
          variant="contained"
          sx={{ mt: 3, mb: 2 }}
          onClick={this.handleSubmit}
          loading={this.state.loadingState === 'loading'}
          disabled={disabled}
        >
          パスワード更新
        </LoadingButton>
      </Box>
    );

    return (
      <AdminWrapper>
        <Box sx={{ mb: 3 }}>
          <BackToNavi text={PAGE_NAMES.accountSettings} link={ACCOUNT_SETTINGS_LINK} />
        </Box>
        <Stack spacing={1} sx={{ px: 1, maxWidth: '400px', mx: 'auto' }}>
          <Typography component="h1" variant="h5" align="center">
            {PAGE_NAMES.passwordChange}
          </Typography>
          {this.renderHeaderInfo()}
          {form}
        </Stack>
      </AdminWrapper>
    );
  }

  protected renderHeaderInfo() {
    const { errorTitle, loadingState } = this.state;

    if (this.props.user?.isSocialAccount) {
      return (
        <Alert severity="info">
          ソーシャルアカウントでログインしています。パスワードを変更することはできません。
        </Alert>
      );
    }

    if (loadingState === 'success') {
      return <Alert>パスワードが変更されました。</Alert>;
    }

    if (errorTitle) {
      return (
        <Box>
          <HeaderErrorMessages errorTitle={errorTitle} />
        </Box>
      );
    }

    return null;
  }

  protected handleChangeTextField =
    (fieldType: 'password1' | 'password2') => (e: React.ChangeEvent<HTMLInputElement>) => {
      switch (fieldType) {
        case 'password1':
          this.setState({
            password1: e.target.value,
          });
          break;
        case 'password2':
          this.setState({
            password2: e.target.value,
          });
          break;
      }
    };

  protected handleSubmit = async () => {
    const { password1, password2 } = this.state;
    this.setState({
      loadingState: 'loading',
      errorTitle: undefined,
      formError: undefined,
    });

    try {
      await autoRefreshApiWrapper(() => changePassword(password1, password2));
      this.setState({
        loadingState: 'success',
        password1: '',
        password2: '',
      });
    } catch (error) {
      const apiError = error as ApiError<ValidationError>;

      this.setState({
        errorTitle: globalAPIErrorMessage(apiError.status, 'submit'),
      });

      if (apiError.status === 400 && apiError.data) {
        // validation Error
        this.setState({
          formError: {
            password1: apiError.data.new_password1,
            password2: apiError.data.new_password2,
          },
        });
      }

      this.setState({
        loadingState: 'error',
      });
    }
  };
}

export type Props = {
  user?: User;
};

export type State = {
  password1: string;
  password2: string;
  loadingState: LoadingState;
  errorTitle?: string;
  formError?: FormError;
};

type FormError = {
  password1?: string[];
  password2?: string[];
};
