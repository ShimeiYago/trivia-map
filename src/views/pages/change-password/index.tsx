import { PAGE_NAMES } from 'constant/page-names';
import { CommonHelmet } from 'helper-components/common-helmet';
import { pageTitleGenerator } from 'utils/page-title-generator';
import { Renderer } from './renderer';

export function ChangePassword() {
  return (
    <>
      <CommonHelmet title={pageTitleGenerator(PAGE_NAMES.passwordChange)} noindex />

      <Renderer />
    </>
  );
}
