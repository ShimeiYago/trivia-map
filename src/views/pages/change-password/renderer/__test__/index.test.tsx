import { shallow, ShallowWrapper } from 'enzyme';
import { Renderer, State } from '..';
import * as ChangePasswordModule from 'api/auths-api/change-password';
import { ApiError } from 'api/utils/handle-axios-error';

let wrapper: ShallowWrapper<unknown, State, Renderer>;

let changePasswordSpy: jest.SpyInstance;

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer />);
  });

  it('basic', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('error case', () => {
    wrapper.setState({
      errorTitle: 'error title',
      formError: {
        password1: ['password is invalid'],
        password2: ['password is invalid'],
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('success case', () => {
    wrapper.setState({
      loadingState: 'success',
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('handleChangeTextField', () => {
  it('should set password1', async () => {
    wrapper = shallow(<Renderer />);
    const instance = wrapper.instance();

    const event = {
      target: {
        value: 'xxxxx',
      },
    } as React.ChangeEvent<HTMLInputElement>;

    instance['handleChangeTextField']('password1')(event);

    expect(instance.state.password1).toBe('xxxxx');
  });

  it('should set password2', async () => {
    wrapper = shallow(<Renderer />);
    const instance = wrapper.instance();

    const event = {
      target: {
        value: 'xxxxx',
      },
    } as React.ChangeEvent<HTMLInputElement>;

    instance['handleChangeTextField']('password2')(event);

    expect(instance.state.password2).toBe('xxxxx');
  });
});

describe('handleSubmit', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    changePasswordSpy = jest.spyOn(ChangePasswordModule, 'changePassword');
  });

  it('should set loadingState success when api succeed', async () => {
    changePasswordSpy.mockResolvedValue({});

    wrapper = shallow(<Renderer />);
    const instance = wrapper.instance();

    await instance['handleSubmit']();

    expect(instance.state.loadingState).toBe('success');
  });

  it('should set form error when api have validation error', async () => {
    const apiError: ApiError<ChangePasswordModule.ValidationError> = {
      status: 400,
      data: {
        password1: ['password is invalid'],
        password2: ['password is invalid'],
      },
      errorMsg: '400 request is invalid',
    };
    changePasswordSpy.mockRejectedValue(apiError);

    wrapper = shallow(<Renderer />);
    const instance = wrapper.instance();

    await instance['handleSubmit']();

    expect(instance.state.formError).toEqual({
      password1: ['password is invalid'],
      password2: ['password is invalid'],
    });
  });

  it('should set loadingstate error when api fail', async () => {
    changePasswordSpy.mockRejectedValue(new Error());

    wrapper = shallow(<Renderer />);
    const instance = wrapper.instance();

    await instance['handleSubmit']();

    expect(instance.state.loadingState).toBe('error');
  });
});
