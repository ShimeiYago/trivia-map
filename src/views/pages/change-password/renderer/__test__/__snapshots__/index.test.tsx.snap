// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests basic 1`] = `
<AdminWrapper>
  <ForwardRef(Box)
    sx={
      Object {
        "mb": 3,
      }
    }
  >
    <BackToNavi
      link="/admin/account"
      text="アカウント管理"
    />
  </ForwardRef(Box)>
  <ForwardRef(Stack)
    spacing={1}
    sx={
      Object {
        "maxWidth": "400px",
        "mx": "auto",
        "px": 1,
      }
    }
  >
    <ForwardRef(Typography)
      align="center"
      component="h1"
      variant="h5"
    >
      パスワード変更
    </ForwardRef(Typography)>
    <ForwardRef(Box)
      component="form"
      noValidate={true}
    >
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        fullWidth={true}
        id="password1"
        label="パスワード"
        margin="normal"
        name="password1"
        onChange={[Function]}
        type="password"
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        fullWidth={true}
        id="password2"
        label="パスワード（確認）"
        margin="normal"
        name="password2"
        onChange={[Function]}
        type="password"
      />
      <ForwardRef(LoadingButton)
        disabled={false}
        fullWidth={true}
        loading={false}
        onClick={[Function]}
        sx={
          Object {
            "mb": 2,
            "mt": 3,
          }
        }
        variant="contained"
      >
        パスワード更新
      </ForwardRef(LoadingButton)>
    </ForwardRef(Box)>
  </ForwardRef(Stack)>
</AdminWrapper>
`;

exports[`Shallow Snapshot Tests error case 1`] = `
<AdminWrapper>
  <ForwardRef(Box)
    sx={
      Object {
        "mb": 3,
      }
    }
  >
    <BackToNavi
      link="/admin/account"
      text="アカウント管理"
    />
  </ForwardRef(Box)>
  <ForwardRef(Stack)
    spacing={1}
    sx={
      Object {
        "maxWidth": "400px",
        "mx": "auto",
        "px": 1,
      }
    }
  >
    <ForwardRef(Typography)
      align="center"
      component="h1"
      variant="h5"
    >
      パスワード変更
    </ForwardRef(Typography)>
    <ForwardRef(Box)>
      <HeaderErrorMessages
        errorTitle="error title"
      />
    </ForwardRef(Box)>
    <ForwardRef(Box)
      component="form"
      noValidate={true}
    >
      <ForwardRef(TextField)
        disabled={false}
        error={true}
        fullWidth={true}
        helperText={
          Array [
            "password is invalid",
          ]
        }
        id="password1"
        label="パスワード"
        margin="normal"
        name="password1"
        onChange={[Function]}
        type="password"
      />
      <ForwardRef(TextField)
        disabled={false}
        error={true}
        fullWidth={true}
        helperText={
          Array [
            "password is invalid",
          ]
        }
        id="password2"
        label="パスワード（確認）"
        margin="normal"
        name="password2"
        onChange={[Function]}
        type="password"
      />
      <ForwardRef(LoadingButton)
        disabled={false}
        fullWidth={true}
        loading={false}
        onClick={[Function]}
        sx={
          Object {
            "mb": 2,
            "mt": 3,
          }
        }
        variant="contained"
      >
        パスワード更新
      </ForwardRef(LoadingButton)>
    </ForwardRef(Box)>
  </ForwardRef(Stack)>
</AdminWrapper>
`;

exports[`Shallow Snapshot Tests success case 1`] = `
<AdminWrapper>
  <ForwardRef(Box)
    sx={
      Object {
        "mb": 3,
      }
    }
  >
    <BackToNavi
      link="/admin/account"
      text="アカウント管理"
    />
  </ForwardRef(Box)>
  <ForwardRef(Stack)
    spacing={1}
    sx={
      Object {
        "maxWidth": "400px",
        "mx": "auto",
        "px": 1,
      }
    }
  >
    <ForwardRef(Typography)
      align="center"
      component="h1"
      variant="h5"
    >
      パスワード変更
    </ForwardRef(Typography)>
    <ForwardRef(Alert)>
      パスワードが変更されました。
    </ForwardRef(Alert)>
    <ForwardRef(Box)
      component="form"
      noValidate={true}
    >
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        fullWidth={true}
        id="password1"
        label="パスワード"
        margin="normal"
        name="password1"
        onChange={[Function]}
        type="password"
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        fullWidth={true}
        id="password2"
        label="パスワード（確認）"
        margin="normal"
        name="password2"
        onChange={[Function]}
        type="password"
      />
      <ForwardRef(LoadingButton)
        disabled={false}
        fullWidth={true}
        loading={false}
        onClick={[Function]}
        sx={
          Object {
            "mb": 2,
            "mt": 3,
          }
        }
        variant="contained"
      >
        パスワード更新
      </ForwardRef(LoadingButton)>
    </ForwardRef(Box)>
  </ForwardRef(Stack)>
</AdminWrapper>
`;
